### PR TITLE
setup terraform

### PR DESCRIPTION
**Pull request by amar-fw**
Plan details

```Terraform will perform the following actions:

  # aws_apigatewayv2_api.api will be created
  + resource "aws_apigatewayv2_api" "api" {
      + api_endpoint                 = (known after apply)
      + api_key_selection_expression = "$request.header.x-api-key"
      + arn                          = (known after apply)
      + execution_arn                = (known after apply)
      + id                           = (known after apply)
      + name                         = "vcc-dev-api"
      + protocol_type                = "HTTP"
      + route_selection_expression   = "$request.method $request.path"
      + tags_all                     = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
    }

  # aws_apigatewayv2_integration.api will be created
  + resource "aws_apigatewayv2_integration" "api" {
      + api_id                                    = (known after apply)
      + connection_type                           = "INTERNET"
      + description                               = "vcc-dev-api"
      + id                                        = (known after apply)
      + integration_method                        = "POST"
      + integration_response_selection_expression = (known after apply)
      + integration_type                          = "AWS_PROXY"
      + integration_uri                           = (known after apply)
      + payload_format_version                    = "1.0"
      + timeout_milliseconds                      = (known after apply)
    }

  # aws_apigatewayv2_route.api will be created
  + resource "aws_apigatewayv2_route" "api" {
      + api_id             = (known after apply)
      + api_key_required   = false
      + authorization_type = "NONE"
      + id                 = (known after apply)
      + route_key          = "$default"
      + target             = (known after apply)
    }

  # aws_apigatewayv2_stage.api will be created
  + resource "aws_apigatewayv2_stage" "api" {
      + api_id        = (known after apply)
      + arn           = (known after apply)
      + auto_deploy   = true
      + deployment_id = (known after apply)
      + execution_arn = (known after apply)
      + id            = (known after apply)
      + invoke_url    = (known after apply)
      + name          = "$default"
      + tags_all      = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }

      + access_log_settings {
          + destination_arn = (known after apply)
          + format          = jsonencode(
                {
                  + errorMessage   = "$context.error.message"
                  + httpMethod     = "$context.httpMethod"
                  + ip             = "$context.identity.sourceIp"
                  + path           = "$context.path"
                  + requestId      = "$context.requestId"
                  + requestTime    = "$context.requestTime"
                  + responseLength = "$context.responseLength"
                  + status         = "$context.status"
                }
            )
        }
    }

  # aws_cloudfront_distribution.app will be created
  + resource "aws_cloudfront_distribution" "app" {
      + arn                            = (known after apply)
      + caller_reference               = (known after apply)
      + comment                        = "vcc-dev-app"
      + default_root_object            = "index.html"
      + domain_name                    = (known after apply)
      + enabled                        = true
      + etag                           = (known after apply)
      + hosted_zone_id                 = (known after apply)
      + http_version                   = "http2"
      + id                             = (known after apply)
      + in_progress_validation_batches = (known after apply)
      + is_ipv6_enabled                = true
      + last_modified_time             = (known after apply)
      + price_class                    = "PriceClass_100"
      + retain_on_delete               = false
      + status                         = (known after apply)
      + tags_all                       = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + trusted_key_groups             = (known after apply)
      + trusted_signers                = (known after apply)
      + wait_for_deployment            = true

      + custom_error_response {
          + error_code         = 404
          + response_code      = 200
          + response_page_path = "/"
        }

      + default_cache_behavior {
          + allowed_methods        = [
              + "GET",
              + "HEAD",
              + "OPTIONS",
            ]
          + cache_policy_id        = "4135ea2d-6df8-44a3-9df3-4b5a84be39ad"
          + cached_methods         = [
              + "GET",
              + "HEAD",
              + "OPTIONS",
            ]
          + compress               = false
          + default_ttl            = (known after apply)
          + max_ttl                = (known after apply)
          + min_ttl                = 0
          + target_origin_id       = "app"
          + trusted_key_groups     = (known after apply)
          + trusted_signers        = (known after apply)
          + viewer_protocol_policy = "redirect-to-https"

          + function_association {
              + event_type   = "viewer-response"
              + function_arn = (known after apply)
            }
        }

      + logging_config {
          + bucket          = (known after apply)
          + include_cookies = false
        }

      + ordered_cache_behavior {
          + allowed_methods        = [
              + "DELETE",
              + "GET",
              + "HEAD",
              + "OPTIONS",
              + "PATCH",
              + "POST",
              + "PUT",
            ]
          + cached_methods         = [
              + "GET",
              + "HEAD",
            ]
          + compress               = true
          + default_ttl            = 0
          + max_ttl                = 0
          + min_ttl                = 0
          + path_pattern           = "/api/*"
          + target_origin_id       = "api"
          + viewer_protocol_policy = "redirect-to-https"

          + forwarded_values {
              + headers                 = [
                  + "authorization",
                  + "user-agent",
                  + "x-forwarded-for",
                ]
              + query_string            = true
              + query_string_cache_keys = (known after apply)

              + cookies {
                  + forward = "all"
                }
            }

          + function_association {
              + event_type   = "viewer-response"
              + function_arn = (known after apply)
            }
        }

      + origin {
          + connection_attempts = 3
          + connection_timeout  = 10
          + domain_name         = (known after apply)
          + origin_id           = "api"

          + custom_origin_config {
              + http_port                = 80
              + https_port               = 443
              + origin_keepalive_timeout = 5
              + origin_protocol_policy   = "https-only"
              + origin_read_timeout      = 30
              + origin_ssl_protocols     = [
                  + "TLSv1.2",
                ]
            }
        }
      + origin {
          + connection_attempts = 3
          + connection_timeout  = 10
          + domain_name         = (known after apply)
          + origin_id           = "app"

          + s3_origin_config {
              + origin_access_identity = (known after apply)
            }
        }

      + restrictions {
          + geo_restriction {
              + locations        = (known after apply)
              + restriction_type = "none"
            }
        }

      + viewer_certificate {
          + cloudfront_default_certificate = true
          + minimum_protocol_version       = "TLSv1"
        }
    }

  # aws_cloudfront_function.response will be created
  + resource "aws_cloudfront_function" "response" {
      + arn             = (known after apply)
      + code            = <<-EOT
            function handler(event) {
              var response = event.response;
              var headers = response.headers;
            
              // Set HTTP security headers
              headers["strict-transport-security"] = {
                value: "max-age=63072000; includeSubdomains; preload",
              };
              headers["content-security-policy"] = {
                value:
                  "default-src 'self'; script-src 'self' 'unsafe-eval' 'unsafe-inline'; style-src 'self' 'unsafe-inline';",
              };
              headers["x-content-type-options"] = { value: "nosniff" };
              headers["x-frame-options"] = { value: "DENY" };
              headers["x-xss-protection"] = { value: "1; mode=block" };
              headers["style-src"] = { value: "'self' 'unsafe-inline'; " };
            
              // Return the response to viewers
              return response;
            }
        EOT
      + comment         = "Add security headers"
      + etag            = (known after apply)
      + id              = (known after apply)
      + live_stage_etag = (known after apply)
      + name            = "vcc-dev-cf-response"
      + publish         = true
      + runtime         = "cloudfront-js-1.0"
      + status          = (known after apply)
    }

  # aws_cloudfront_origin_access_identity.app will be created
  + resource "aws_cloudfront_origin_access_identity" "app" {
      + caller_reference                = (known after apply)
      + cloudfront_access_identity_path = (known after apply)
      + comment                         = "vcc-dev-app"
      + etag                            = (known after apply)
      + iam_arn                         = (known after apply)
      + id                              = (known after apply)
      + s3_canonical_user_id            = (known after apply)
    }

  # aws_cloudwatch_log_group.api_gateway will be created
  + resource "aws_cloudwatch_log_group" "api_gateway" {
      + arn               = (known after apply)
      + id                = (known after apply)
      + name              = "api-gateway/vcc-dev-api"
      + retention_in_days = 0
      + tags_all          = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
    }

  # aws_iam_role.lambda will be created
  + resource "aws_iam_role" "lambda" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = [
                              + "lambda.amazonaws.com",
                            ]
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + description           = "Lambda execution role for vcc-dev"
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "vcc-dev-lambda"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_role_policy_attachment.lambda_cloudwatch will be created
  + resource "aws_iam_role_policy_attachment" "lambda_cloudwatch" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/CloudWatchFullAccess"
      + role       = "vcc-dev-lambda"
    }

  # aws_iam_role_policy_attachment.lambda_dynamodb will be created
  + resource "aws_iam_role_policy_attachment" "lambda_dynamodb" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess"
      + role       = "vcc-dev-lambda"
    }

  # aws_iam_role_policy_attachment.lambda_execute will be created
  + resource "aws_iam_role_policy_attachment" "lambda_execute" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaRole"
      + role       = "vcc-dev-lambda"
    }

  # aws_iam_role_policy_attachment.lambda_s3 will be created
  + resource "aws_iam_role_policy_attachment" "lambda_s3" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/AmazonS3FullAccess"
      + role       = "vcc-dev-lambda"
    }

  # aws_iam_role_policy_attachment.lambda_sqs will be created
  + resource "aws_iam_role_policy_attachment" "lambda_sqs" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/AmazonSQSFullAccess"
      + role       = "vcc-dev-lambda"
    }

  # aws_iam_role_policy_attachment.lambda_vpc will be created
  + resource "aws_iam_role_policy_attachment" "lambda_vpc" {
      + id         = (known after apply)
      + policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
      + role       = "vcc-dev-lambda"
    }

  # aws_lambda_function.api will be created
  + resource "aws_lambda_function" "api" {
      + architectures                  = (known after apply)
      + arn                            = (known after apply)
      + description                    = "API for vcc-dev"
      + filename                       = "build/api.zip"
      + function_name                  = "vcc-dev-api"
      + handler                        = "lambda.handler"
      + id                             = (known after apply)
      + invoke_arn                     = (known after apply)
      + last_modified                  = (known after apply)
      + memory_size                    = 2048
      + package_type                   = "Zip"
      + publish                        = false
      + qualified_arn                  = (known after apply)
      + reserved_concurrent_executions = -1
      + role                           = (known after apply)
      + runtime                        = "nodejs14.x"
      + signing_job_arn                = (known after apply)
      + signing_profile_version_arn    = (known after apply)
      + source_code_hash               = "aB9ZX+kMU8ZwxZ/jLxglwd4FGVhDJ1kSkrNAMoUMhlc="
      + source_code_size               = (known after apply)
      + tags_all                       = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + timeout                        = 30
      + version                        = (known after apply)

      + environment {
          + variables = {
              + "NODE_ENV" = "dev"
            }
        }

      + tracing_config {
          + mode = (known after apply)
        }
    }

  # aws_lambda_permission.api_allow_gateway will be created
  + resource "aws_lambda_permission" "api_allow_gateway" {
      + action        = "lambda:InvokeFunction"
      + function_name = "vcc-dev-api"
      + id            = (known after apply)
      + principal     = "apigateway.amazonaws.com"
      + source_arn    = (known after apply)
      + statement_id  = (known after apply)
    }

  # aws_s3_bucket.app will be created
  + resource "aws_s3_bucket" "app" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "vcc-dev-app"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + logging {
          + target_bucket = (known after apply)
          + target_prefix = "vcc-dev-app/"
        }

      + versioning {
          + enabled    = true
          + mfa_delete = false
        }
    }

  # aws_s3_bucket.app_logs will be created
  + resource "aws_s3_bucket" "app_logs" {
      + acceleration_status         = (known after apply)
      + acl                         = "log-delivery-write"
      + arn                         = (known after apply)
      + bucket                      = "vcc-dev-app-logs"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }
    }

  # aws_s3_bucket.bucket_logs will be created
  + resource "aws_s3_bucket" "bucket_logs" {
      + acceleration_status         = (known after apply)
      + acl                         = "log-delivery-write"
      + arn                         = (known after apply)
      + bucket                      = "vcc-dev-bucket-logs"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = {
          + "Author"      = "FreshWorks Studio"
          + "Environment" = "dev"
          + "Project"     = "vcc"
        }
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }
    }

  # aws_s3_bucket_policy.app will be created
  + resource "aws_s3_bucket_policy" "app" {
      + bucket = (known after apply)
      + id     = (known after apply)
      + policy = (known after apply)
    }

Plan: 21 to add, 0 to change, 0 to destroy.```

### CODE REVIEW

File: terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,43 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "3.74.2"
```
Comment (will-freshworks) [inline terraform/main.tf]: Latest version is `4.8.0` and I think we should use the latest one

File: terraform/iam.tf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  role       = aws_iam_role.lambda.name
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaRole"
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_s3" {
+  role       = aws_iam_role.lambda.name
+  policy_arn = "arn:aws:iam::aws:policy/AmazonS3FullAccess"
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_vpc" {
```
Comment (will-freshworks) [inline terraform/iam.tf]: Not using a VPC so we can skip this

File: terraform/iam.tf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    ]
+  })
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_cloudwatch" {
+  role       = aws_iam_role.lambda.name
+  policy_arn = "arn:aws:iam::aws:policy/CloudWatchFullAccess"
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_sqs" {
```
Comment (will-freshworks) [inline terraform/iam.tf:25]: Not using SQS but who knows for the future so we could leave this in

**Review by will-freshworks — DISMISSED** [review]
Minor changes. Good to go ahead and merge once they are made. 

Btw have you tested this and do we have a cloudfront URL for dev already?

**Comment by amar-fw** [comment]
> Btw have you tested this and do we have a cloudfront URL for dev already?

@will-freshworks  Nope, I just ran a terraform plan. I think we need a domain name for the CloudFront url right? 🤔

**Pull request merged**